// @sentry/wizard/dist/lib/Helper/Logging.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.debug = exports.cyan = exports.yellow = exports.dim = exports.red = exports.green = exports.nl = exports.l = exports.prepareMessage = void 0;
const chalk_1 = __importDefault(require("chalk"));
function prepareMessage(msg) {
    if (typeof msg === 'string') {
        return msg;
    }
    if (msg instanceof Error) {
        return `${msg.stack || ''}`;
    }
    return JSON.stringify(msg, null, '\t');
}
exports.prepareMessage = prepareMessage;
function l(msg) {
    // eslint-disable-next-line no-console
    console.log(msg);
}
exports.l = l;
function nl() {
    return l('');
}
exports.nl = nl;
function green(msg) {
    return l(chalk_1.default.green(prepareMessage(msg)));
}
exports.green = green;
function red(msg) {
    return l(chalk_1.default.red(prepareMessage(msg)));
}
exports.red = red;
function dim(msg) {
    return l(chalk_1.default.dim(prepareMessage(msg)));
}
exports.dim = dim;
function yellow(msg) {
    return l(chalk_1.default.yellow(prepareMessage(msg)));
}
exports.yellow = yellow;
function cyan(msg) {
    return l(chalk_1.default.cyan(prepareMessage(msg)));
}
exports.cyan = cyan;
/**
 * @deprecated Use `debug` from `src/utils/debug.ts` instead.
 */
function debug(msg) {
    return l(chalk_1.default.italic.yellow(prepareMessage(msg)));
}
exports.debug = debug;
